lib/lf/assembly/dofhandler.h

```diff
@@ -0,0 +1,187 @@
+/***************************************************************************
+ * LehrFEM++ - A simple C++ finite element libray for teaching
+ * Developed from 2018 at the Seminar of Applied Mathematics of ETH Zurich,
+ * lead developers Dr. R. Casagrande and Prof. R. Hiptmair
+ ***************************************************************************/
+
+/**
+ * @file
+ * @brief A general DOF handler interface
+ * @author Ralf Hiptmair
+ * @date August 2018
+ * @copyright MIT License
+ */
+
+#ifndef _LF_DOFHD_H
+#define _LF_DOFHD_H
+
+#include "local_static_dofs.h"
+
+namespace lf::assemble {
+/*
+ * @brief A general (_interface_) class for DOF handling.
+ *
+ * Objects of this class provide the local-to-global map for indices
+ * of local/global shape functions. 
+ */
+class DofHandler {
+ public:
+  /**
+   * @brief The constructor just stores the mesh pointer
+   */
+  DofHandler(void) {}
+  virtual ~DofHandler(void) = default;
+
+  DofHandler(const DofHandler &) = delete;
+  DofHandler(DofHandler &&) = delete;
+  DofHandler &operator=(const DofHandler &) = delete;
+  DofHandler &operator=(DofHandler &&) = delete;
+  /**
+   * @brief total number of dof's handled by the object
+   */
+  virtual size_type GetNoDofs(void) const = 0;
+
+  /**
+   * @brief access to indices of global dof's belonging to an entity
+   *
+   * @param ref_el_type type of entity for which to retrieve dof indices
+   * @param entity_index unique mesh index for an entity of the specified type
+   *
+   * The basis functions of every finite element space must be associated with a
+   * unique geometric entity. Conversely, every entity can possess a finite
+   * number of basis functions = degrees of freedom. This functions return the
+   * global indices of all basis functions associated with the entity _and its
+   * sub-entitites_.
```

Comment:
I have already contemplated adding a method for fetching "interior" degrees of freedom. Will be available in the final release of the assembly module.

Reply:
Well, I have changed my mind, because the user of assembly facilities must have a very clear idea about the numbering of local shape functions. Thus, he knows precisely, which local index numbers refer to interior local shape functions.